### ISSUE TITLE
[13.0] framework: XML-RPC calls raise ValidationError with sql_constraints even if no input

### ISSUE THREAD

**Issue by Yenthe666** [odoo/odoo#41438]
Impacted versions: 13.0 atleast

Steps to reproduce: Let's go for a long ride to test this. Steady your horses. 🐴 
1. Create a new module, inherit product.template and create a new `_sql_constraints` item:
```
_sql_constraints = [
        ('product_template_default_code_uniq', 'unique(default_code)',
         "The default code has to be unique and can only be assigned to one product!")
    ]
```
This constraint will make sure that if a user enters the same `default_code` twice that he/she gets a `ValidationError`.
2. Install the custom module you've built
3. Create a new product "ABC1234" with "123456789" as `default_code`
4. Create a second product "ABC12345" with no `default_code`
5. Edit the second product again and fill in the `default_code` "123456789" you'll now get the constraint error as expected:
![image](https://user-images.githubusercontent.com/6352350/70263867-96756d00-1797-11ea-9635-e3beb485d59d.png)
Just remove the `default_code` again and save.  Notice how your product will be succesfully saved (as expected).
6. Now create an XML-RPC script that searches the product "ABC1234" and let us remove the `default_code` from this record. The idea is to update products in bulk through XML-RPC and updating the codes. Here's an XML-RPC script to test with:
```
import xmlrpc.client

url = 'https://your-odoo-instance.com'
db = 'development'
username = 'yourusername'
password = 'yourpassword'


common = xmlrpc.client.ServerProxy('{}/xmlrpc/2/common'.format(url))
uid = common.authenticate(db, username, password, {})
models = xmlrpc.client.ServerProxy('{}/xmlrpc/2/object'.format(url))
product_id = models.execute_kw(db, uid, password, 'product.template', 'search', [[['name', '=', 'ABC1234']]])
if not product_id:
    product_id = models.execute_kw(db, uid, password, 'product.template', 'create' , [{
      'name': 'ABC1234',
      'default_code': '123456789'
    }])
else:
    product_id = product_id[0]
    print('product_id: ' + str(product_id))
    models.execute_kw(db, uid, password, 'product.template', 'write', [[product_id], {
        'default_code': ''
    }])

```

Current behavior:
You'll get a traceback:
```
xmlrpc.client.Fault: <Fault 1: 'Traceback (most recent call last):\n  File "/odoo13/odoo13-server/odoo/service/model.py", line 93, in wrapper\n    return f(dbname, *args, **kwargs)\n  File "/odoo13/odoo13-server/odoo/service/model.py", line 175, in execute\n    res = execute_cr(cr, uid, obj, method, *args, **kw)\n  File "/odoo13/odoo13-server/odoo/service/model.py", line 164, in execute_cr\n    return odoo.api.call_kw(recs, method, args, kw)\n  File "/odoo13/odoo13-server/odoo/api.py", line 396, in call_kw\n    model.flush()\n  File "/odoo13/odoo13-server/odoo/models.py", line 5346, in flush\n    process(self.env[model_name], id_vals)\n  File "/odoo13/odoo13-server/odoo/models.py", line 5337, in process\n    recs._write(vals)\n  File "/odoo13/odoo13-server/odoo/models.py", line 3593, in _write\n    cr.execute(query, params + [sub_ids])\n  File "/odoo13/odoo13-server/odoo/sql_db.py", line 163, in wrapper\n    return f(self, *args, **kwargs)\n  File "/odoo13/odoo13-server/odoo/sql_db.py", line 240, in execute\n    res = self._obj.execute(query, params)\npsycopg2.IntegrityError: duplicate key value violates unique constraint "product_template_product_template_default_code_uniq"\nDETAIL:  Key (default_code)=() already exists.\n\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File "/odoo13/odoo13-server/odoo/addons/base/controllers/rpc.py", line 63, in xmlrpc_2\n    response = self._xmlrpc(service)\n  File "/odoo13/odoo13-server/odoo/addons/base/controllers/rpc.py", line 43, in _xmlrpc\n    result = dispatch_rpc(service, method, params)\n  File "/odoo13/odoo13-server/odoo/http.py", line 138, in dispatch_rpc\n    result = dispatch(method, params)\n  File "/odoo13/odoo13-server/odoo/service/model.py", line 40, in dispatch\n    res = fn(db, uid, *params)\n  File "/odoo13/odoo13-server/odoo/service/model.py", line 168, in execute_kw\n    return execute(db, uid, obj, method, *args, **kw or {})\n  File "/odoo13/odoo13-server/odoo/service/model.py", line 115, in wrapper\n    raise ValidationError(tr(key, \'sql_constraint\') or inst.pgerror)\nodoo.exceptions.ValidationError: (\'product_template_product_template_default_code_uniq\', None)\n'>
```

Expected behavior:
There shouldn't be any traceback. Our product "ABC1234" had the `default_code` "123456789" and was the only product in the database with this `default_code`. Now that we do a write on this record we'll set it to "", which should be ignored by the constraint. When you create two products in Odoo, through the frontend, with no `default_code` it will not give a constrain error as empty values are ignored. When you do the same through XML-RPC the `ValidationError` is triggered, even when the value is empty.
If you'd change the XML-RPC example script and set the `default_code` to a random value (for example "555555") it will work just fine too.

In conclusion: XML-RPC does not seem to ignore `_sql_constraints` when the value for the field is empty (`''`) and will still give back the constraint error while it shouldn't go off.

**Comment by pedrobaeza** [comment]
Why don't you put `'default_code': False`?

**Comment by Yenthe666** [comment]
Euhh I tested both of them (both cases fail) and to me '' was more logical as it is a char field! Any reason why you'd prefer False?

**Comment by pedrobaeza** [comment]
Well, all fields are nullified through `False`, and I remember some inconsistency with NULL/'' in database.

**Comment by Yenthe666** [comment]
Hmm, doing `False` instead of `''` seems to work in some cases actually. @pedrobaeza do you usually set char fields as `False` in XML-RPC then?

**Comment by pedrobaeza** [comment]
I'm not using XML-RPC lately, but it was just I remember that kind of things when I use to do it. There are 2 factors here:

- The XML-RPC protocol itself, that doesn't allow for example to transfer `None` values.
- The Odoo bridge from XML-RPC to ORM.

**Comment by Yenthe666** [comment]
I've also find a side-effect with barcodes and archived products so let's wait for the OPW but I appreciate your feedback @pedrobaeza :)

**Comment by nle-odoo** [comment]
The fact that in database we can have empty string or NULL that are different falsy values has been a fact for a long time.

In a 9.0 shell:

```
>>> self.env['res.users'].browse(3).email = ''
>>> self.env['res.users'].browse(3).email
u''
>>> self.env['res.users'].browse(3).email = None
>>> self.env['res.users'].browse(3).email
False
```

In a 13.0 shell:

```
In [1]: self.env['res.users'].browse(3).email = ''                                                                                                                                                                                             

In [2]: self.env['res.users'].browse(3).email                                                                                                                                                                                                  
Out[2]: ''

In [3]: self.env['res.users'].browse(3).email = None                                                                                                                                                                                           

In [4]: self.env['res.users'].browse(3).email                                                                                                                                                                                                  
Out[4]: False
```

The javascript framework will always send `false` (at least since 9.0,) so by normal usage writing empty string in database should not happen.

But if you use `write` yourself (by RPC or code) you can save the empty string `''`.

Since people could expect the behavior to exist in custom code, this is not something we should change in stable.

**Comment by Yenthe666** [comment]
Thanks for the feedback @nle-odoo. Do I get you right that you recommend to always send `False` through XML-RPC instead of `''` then?

**Comment by nle-odoo** [comment]
Yes, `False` or `None`.

As a side note, this is not related to RPC, if you do in python code:

record.textfield = ''

the behavior is similar.

**Comment by Yenthe666** [comment]
Resolved in https://github.com/odoo/odoo/commit/cd412eb0d13b9c615
